### PR TITLE
WIP: Validate generated Pydantic models on Synthea.

### PR DESCRIPTION
**Pull request by axelv**
First of all thx for starting this awesome library!

This is WIP but I'd like your feedback on how to proceed. My goal is to validate the generated models on [Synthea generated FHIR resources](https://synthea.mitre.org/downloads) so we that we are confident that the models work correctly. During this attempt I encountered several issues:

1. Pydantic models have circular references which need to be resolved using [`update_forward_refs`](https://docs.pydantic.dev/usage/postponed_annotations/). I added this in `ast.py`

2. We will need a [union type of all resources](https://docs.pydantic.dev/usage/types/#unions) to parse resources inside a Bundle. Currently, Pydantic tries to validate `Bundle.entry.resource` as a `Resource` but actually, it should validate it as a Union of all possible resources. I added the automatic generation of such a union type (see: `AnyResource`) to `ast.py`.

3. `resourceType` was missing. I tried to generate it automatically but failed to do it correctly for all resources because some resources are split into multiple models. I think this is the result of polymorphic elements. Ex `Immunization = _ImmunizationDateTime | _ImmunizationString`. Having two models with the same `resourceType` is not practical because then we can't use [Pydantic's `Field(discriminator='resourceType')` for Union types](https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions)

ToDo:
- fix tests according to the changes
- fix 'resourceType' 

Any thoughts on the issues above?
Feedback on code style and directory structure is welcome!

### CODE REVIEW

File: fhir_py_types/ast.py [hunk 1]
```diff
@@ -248,12 +248,25 @@ def iterate_definitions_tree(
 
     yield root
 
+def update_forward_refs(definition: StructureDefinition) -> Iterable[ast.stmt]:
```
Comment (axelv) [inline fhir_py_types/ast.py]: Ex. for StructureDefinition of "Account":
`Account.update_forward_refs()`

File: fhir_py_types/ast.py [hunk 2]
```diff
@@ -262,7 +275,7 @@ def build_ast(
                         typedefinitions.extend(define_polymorphic(definition))
                     else:
                         typedefinitions.extend(define_class(definition))
-
+                    forward_references.extend(update_forward_refs(definition))
```
Comment (axelv) [inline fhir_py_types/ast.py]: add forward references for all complex types and resources

File: fhir_py_types/ast.py [hunk 3]
```diff
@@ -275,7 +275,7 @@ def build_ast(
                         typedefinitions.extend(define_polymorphic(definition))
                     else:
                         typedefinitions.extend(define_class(definition))
-                    forward_references.extend(update_forward_refs(definition))
+                        forward_references.extend(update_forward_refs(definition))
```
Comment (axelv) [inline fhir_py_types/ast.py]: Union types don't need forward refs.

File: fhir_py_types/ast.py [hunk 4]
```diff
@@ -28,11 +28,14 @@ class AnnotationForm(Enum):
 def make_type_annotation(
     type_: StructurePropertyType, form: AnnotationForm
 ) -> ast.expr:
+    name = type_.code
+    if name == "Resource":
+        name = "AnyResource"
```
Comment (axelv) [inline fhir_py_types/ast.py]: Workaround to replace "Resource" with the union of all resources which is "AnyResource"

File: fhir_py_types/ast.py [hunk 5]
```diff
@@ -145,18 +148,23 @@ def define_class_object(
...
+    body_stmts = [*itertools.chain.from_iterable(
+                    type_annotate(property, identifier, AnnotationForm.Property)
+                    for identifier, property in order_type_overriding_properties(
+                        definition.elements
+                    )
+                )]
+    if definition.kind == StructureDefinitionKind.RESOURCE:
+        annotation = ast.Subscript(ast.Name("Literal"), slice=ast.Str(definition.id))
+        body_stmts.insert(0, ast.AnnAssign(target=[ast.Name("resourceType")], annotation=annotation, simple=1))
+
```
Comment (axelv) [inline fhir_py_types/ast.py]: if kind="resource" then add "resourceType" as an extra element to the model

File: fhir_py_types/ast.py [hunk 6]
```diff
@@ -248,6 +256,29 @@ def iterate_definitions_tree(
 
     yield root
 
+def has_resource_type(definition: StructureDefinition) -> bool:
+    return "resourceType" in definition.elements
+
+def define_any_resource(resource_definitions:set[StructureDefinition]) -> ast.stmt:
```
Comment (axelv) [inline fhir_py_types/ast.py]: create a union of all resources with the name "AnyResource"

**Comment by m0rl** [comment]
Thanks for helping us making this library even better! There were some updates to the main branch recently that include `resourceType` generation and automation of `update_forward_refs` for `BaseModel` resources. Please let us know if there are any issues left with that. As for the `AnyResource` type - I believe it's a great idea and we can also extend it to other contexts where we now have `Resource` (like `contained` property). Will look into it!

**Comment by m0rl** [comment]
Quick update on the proposed changes: synthea-generated resources validation should now work in the main branch.

**Comment by axelv** [comment]
Very cool @m0rl! Meanwhile I'm looking for an elegant way to implement the Any* types for abstract structures like Resource, DomainResource. Not sure if this also holds for Element and BackboneElement (these also have `abstract = true`).

**Comment by axelv** [comment]
> Very cool @m0rl! Meanwhile I'm looking for an elegant way to implement the Any* types for abstract structures like Resource, DomainResource. Not sure if this also holds for Element and BackboneElement (these also have `abstract = true`).

Nevermind. I just notice that you already fixed the AnyResource generation. 
Good job `Bundle.parse_file` on the Synthea works fine👍  🚀 

Pytest results on 1180 Bundles
```
Total number of tests expected to run: 1180
Total number of tests run: 1180
Total number of tests passed: 1180
Total number of tests failed: 0
Total number of tests failed with errors: 0
Total number of tests skipped: 0
Total number of tests with no result data: 0
Finished running tests!
```

**Comment by m0rl** [comment]
Thanks for testing it! We also plan to add synthea bundle validation regression tests in the near future (together with the regression tests for R4/R5 spec parsing and typecheck).

**Pull request closed without merging**